### PR TITLE
[udp-ping-pong] Refactor to add Drop

### LINKED ISSUE #717: [tests] Refactor `udp-ping-pong` to Follow "Associated Functions" Design
## Context

The `udp-ping-pong` application was written as a collection of standalone functions. 

We should refactor the `udp-ping-pong` application to follow an "associated functions" code structure, so that we can have correct trait implementations, such as `Drop`.

### PR DESCRIPTION
**Pull request by anandbonde**
Closes #717

### CODE REVIEW

File: examples/rust/udp-ping-pong.rs [hunk 1]
```diff
@@ -102,240 +103,213 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
...
+            };
+
+            // Sanity check received data.
+            let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+            let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+            for x in slice {
+                if *x != fill_char {
+                    // If not matching, free scatter-gather array.
+                    freesga(&mut self.libos, sga);
```
Comment (iyzhang) [inline examples/rust/udp-ping-pong.rs:150]: Lets do the same and add Option<demi_sgarray_t> to the server struct, so we can get rid of this line and lines 160, 168 and 173.

File: examples/rust/udp-ping-pong.rs [hunk 2]
```diff
@@ -102,240 +103,213 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
...
+
+        // Push and pop data.
+        let sga: demi_sgarray_t = mksga(&mut self.libos, buffer_size, fill_char)?;
+        match self.libos.pushto(self.sockqd, &sga, remote) {
+            Ok(qt) => qts.push(qt),
             Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e)
+                // If error, free scatter-gather array.
```
Comment (iyzhang) [inline examples/rust/udp-ping-pong.rs:233]: Lets do the same here as well.

**Review by iyzhang — APPROVED** [review]
LGTM

File: examples/rust/udp-ping-pong.rs (hunk 2 above)
Comment (anandbonde) [reply examples/rust/udp-ping-pong.rs]: Done.

File: examples/rust/udp-ping-pong.rs (hunk 1 above)
Comment (anandbonde) [reply examples/rust/udp-ping-pong.rs]: Done.

**Pull request merged**